exp/clusterresourceset/controllers/clusterresourceset_controller.go

```diff
@@ -0,0 +1,357 @@
+			r.Log.Error(err, "failed to patch config")
+		}
+	}()
+
+	// Initializes the maps in ClusterResourceSetBinding object if they are uninitialized.
+	InitClusterResourceSetBinding(clusterResourceSetBinding, clusterResourceSet)
+
+	var clusterResourceSetInMap clusterresourcesetv1.ResourcesInClusterResourceSet
+	var ok bool
+	if clusterResourceSetInMap, ok = clusterResourceSetBinding.ClusterResourceSetMap[clusterResourceSet.Name]; !ok {
+		// This should never be the case as InitClusterResourceSetBinding is adding the ClusterResourceSet to the map during initialization.
+		return errors.New("failed to find ClusterResourceSet " + clusterResourceSet.Name + " in ClusterResourceBinding")
+	}
+
+	// Iterate all resources and apply them to the cluster and update the resource status in the ClusterResourceSetBinding object.
+	for _, resource := range clusterResourceSet.Spec.Resources {
+		// If resource is already applied successfully and clusterResourceSet mod is "ApplyOnce", continue. (No need to check hash changes here)
+		resourceInMap := clusterresourcesetv1.ClusterResourceSetResourceStatus{}
+		if resourceInMap, ok = clusterResourceSetInMap.Resources[GenerateResourceKindName(resource)]; ok {
+			if clusterResourceSet.Spec.Mode == string(clusterresourcesetv1.ClusterResourceSetModeApplyOnce) && resourceInMap.Successful {
+				continue
+			}
+		}
+
+		// Only allow using resources (Secrets/Configmaps) in the same namespace with the cluster.
+		typedName := types.NamespacedName{Name: resource.Name, Namespace: cluster.Namespace}
+		dataList := make([][]byte, 0)
+
+		// Retrieve data in the resource as an array because there can be many key-value pairs in resource and all will be applied to the cluster.
+		switch resource.Kind {
+		case string(clusterresourcesetv1.ConfigMapClusterResourceSetResourceKind):
+			resourceConfigMap, err := GetConfigMapFromNamespacedName(context.Background(), r.Client, typedName)
+			if err != nil {
+				return errors.Wrapf(err,
+					"failed to fetch ClusterResourceSet configmap %q in namespace %q", resource.Name, clusterResourceSet.Namespace)
+			}
+
+			for _, dataStr := range resourceConfigMap.Data {
+				data := []byte(dataStr)
+				dataList = append(dataList, data)
+			}
+		case string(clusterresourcesetv1.SecretClusterResourceSetResourceKind):
+			resourceSecret, err := secret.GetAnySecretFromNamespacedName(context.Background(), r.Client, typedName)
+			if err != nil {
+				return errors.Wrapf(err,
+					"failed to fetch ClusterResourceSet secret %q in namespace %q", resource.Name, clusterResourceSet.Namespace)
+			}
+
+			if resourceSecret.Type != clusterresourcesetv1.ClusterResourceSetSecretType {
+				return errors.Wrapf(err,
+					"Unsupported ClusterResourceSet secret resource: %q/%q type: %q", clusterResourceSet.Namespace, resource.Name, resourceSecret.Type)
+			}
+			for _, dataStr := range resourceSecret.Data {
+				dataList = append(dataList, dataStr)
+			}
+
+		default:
```

Comment:
This should never happen due to the validator web hook

Reply:
Right, but in case new modes are added and we forget to validate it in webhook, I'd prefer to keep this.